File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        #BST Solution
        #because its bst, look for node in the middle, starting from root
        if root is None:
            return None

        #found middle value -> LCA
        if min(p.val, q.val) <= root.val and root.val <= max(p.val, q.val):
            return root
        
        #otherwise, search LST/RST
        return self.lowestCommonAncestor(root.left, p, q) or self.lowestCommonAncestor(root.right, p, q)
        

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public class Tuple{
        public int depth;
        public bool isBalanced;

        public Tuple(int depth, bool isBalanced)
        {
            this.depth = depth;
            this.isBalanced = isBalanced;
        }
    }

    public bool IsBalanced(TreeNode root) {
        return DFS(root).isBalanced;
    }

    public Tuple DFS(TreeNode curr){
        //Go through every node bottom up -> DFS POST ORDER
        if(curr == null){
            return new Tuple(0, true);
        }

        Tuple leftVal = DFS(curr.left);
        Tuple rightVal = DFS(curr.right);
        
        bool isBalanced = leftVal.isBalanced && rightVal.isBalanced
                          && Math.Abs(leftVal.depth - rightVal.depth) <= 1;
        
        return new Tuple(1 + Math.Max(leftVal.depth,rightVal.depth), isBalanced);
    }

    
}
